# 🧠 Task 4: Claim Risk & Pricing Models


## 📌 Objective
Build and evaluate predictive models that support risk-based premium pricing for an insurance company.
    

In [1]:
print(7)

7


## ✅ 1. Data Preparation
---

In [2]:
import sys
import os

# Add the root project directory to sys.path
sys.path.append(os.path.abspath(os.path.join(os.getcwd(), '..')))

In [3]:
# load data
import pandas as pd
df = pd.read_csv('../data/processed/cleaned_insurance_data.csv')

C:\Users\Mesfin\AppData\Local\Temp\ipykernel_7400\1886200167.py:3: DtypeWarning: Columns (32) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/processed/cleaned_insurance_data.csv')


In [ ]:
df.info()

## 🔍 2. Claim Severity Regression (TotalClaims > 0)
---

In [4]:
# 👉 Train & evaluate Linear Regression, Random Forest, XGBoost
from src.models.train_claim_severity import train_and_evaluate_claim_severity_models

fitted_pipes_claim_severity = train_and_evaluate_claim_severity_models(df)

📊 Linear Regression
✅ RMSE: 0.28
✅ R²:   1.0000
----------------------------------------
📊 Decision Tree
✅ RMSE: 924.49
✅ R²:   0.9994
----------------------------------------
📊 Random Forest
✅ RMSE: 703.46
✅ R²:   0.9997
----------------------------------------
📊 XGBoost
✅ RMSE: 3817.73
✅ R²:   0.9900
----------------------------------------


## 💰 3. Premium Prediction (CalculatedPremiumPerTerm)
---

In [4]:
# 👉 Train & evaluate models to predict premiums
from src.models.train_premium import train_and_evaluate_premium_models
train_and_evaluate_premium_models(df)

📊 Linear Regression
✅ RMSE: 39.69
✅ R²:   0.9678
----------------------------------------
📊 Decision Tree
✅ RMSE: 22.92
✅ R²:   0.9893
----------------------------------------
📊 Random Forest
✅ RMSE: 22.77
✅ R²:   0.9894
----------------------------------------
📊 XGBoost
✅ RMSE: 25.83
✅ R²:   0.9864
----------------------------------------


## ⚠️ 4. Claim Occurrence Probability (Binary Classification)
---

In [5]:
# 👉 Logistic Regression, Random Forest, XGBoost Classifier
from src.models.train_claim_probability import train_and_evaluate_claim_probability_models
train_and_evaluate_claim_probability_models(df)

📊 Logistic Regression
✅ Accuracy:  1.0000
✅ F1 Score:  1.0000
✅ ROC AUC:   1.0000
----------------------------------------
📊 Decision Tree
✅ Accuracy:  1.0000
✅ F1 Score:  1.0000
✅ ROC AUC:   1.0000
----------------------------------------
📊 Random Forest
✅ Accuracy:  0.9999
✅ F1 Score:  0.9920
✅ ROC AUC:   1.0000
----------------------------------------


c:\ME\KAIM\acis-insurance-risk-analytics\week3\Lib\site-packages\xgboost\training.py:183: UserWarning: [16:16:59] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


📊 XGBoost
✅ Accuracy:  1.0000
✅ F1 Score:  1.0000
✅ ROC AUC:   1.0000
----------------------------------------


## 📊 5. Evaluation Summary
---


#### ✅  Risk-Based Pricing System


1️⃣ Claim Severity Prediction (Regression – `TotalClaims`)

| Model             | RMSE     ↓ | R²       ↑ | Notes                              |
| ----------------- | ---------- | ---------- | ---------------------------------- |
| Linear Regression | **0.28**   | **1.0000** | ⚠️ Too perfect – check for leakage |
| Decision Tree     | 924.49     | 0.9994     | Strong, possibly overfitting       |
| Random Forest     | **703.46** | **0.9997** | ✅ Best balance of accuracy         |
| XGBoost           | 3817.73    | 0.9900     | Underperforming – needs tuning     |



2️⃣ Premium Prediction (Regression – `CalculatedPremiumPerTerm`)

| Model             | RMSE     ↓ | R²       ↑ | Notes                |
| ----------------- | ---------- | ---------- | -------------------- |
| Linear Regression | 39.69      | 0.9678     | Solid baseline       |
| Decision Tree     | 22.92      | 0.9893     | Strong and fast      |
| Random Forest     | **22.77**  | **0.9894** | ✅ Best overall model |
| XGBoost           | 25.83      | 0.9864     | Slightly behind RF   |



3️⃣ Claim Occurrence Prediction (Classification – `ClaimMade`)

| Model               | Accuracy ↑ | F1 Score ↑ | ROC AUC ↑ | Notes                              |
| ------------------- | ---------- | ---------- | --------- | ---------------------------------- |
| Logistic Regression | 1.0000     | 1.0000     | 1.0000    | ⚠️ Too perfect – check for leakage |
| Decision Tree       | 1.0000     | 1.0000     | 1.0000    | Same — possibly overfit            |
| Random Forest       | 0.9999     | 0.9920     | 1.0000    | ✅ Very strong, realistic results   |
| XGBoost             | 1.0000     | 1.0000     | 1.0000    | ⚠️ Possibly overfitting            |



### 🧠 Interpretation

* **Claim Severity** and **Premium** prediction are best handled by **Random Forest**.
* **Claim Occurrence** models are reporting **perfect metrics** — this may indicate:

  * Data leakage (i.e., target accidentally used as input)
  * Or a **very imbalanced dataset** where class 1 dominates
* **XGBoost** is slightly underperforming — may need custom hyperparameters or tuning.

## 🧠 6. Interpretation (SHAP / Feature Importance)
---

In [ ]:
# 👉 SHAP plots, feature importance charts

In [ ]:

# 🔍 Step 1: Extract model, preprocessor, and transformed data (example: random forest from claim severity)
model = pipe.named_steps["regressor"]
X_transformed = pipe.named_steps["preprocessor"].transform(X_train)
feature_names = pipe.named_steps["preprocessor"].get_feature_names_out()
# 🔍 Step 2: Calculate SHAP values

## 📈 7. Final Insights & Recommendation
---